File: April 2022/April_29.cpp
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<int> grp(n, -1);
        for(int i = 0; i < n; i++){
            if(grp[i] == -1){
                queue<int> q;
                q.push(i);
                int g = 0;
                while(!q.empty()){
                    int size = q.size();
                    while(size--){
                        int rem = q.front(); q.pop();
                        if(grp[rem] != -1){
                            if(grp[rem] == g)
                                continue;
                            return false;
                        }
                        grp[rem] = g;
                        for(auto &nbr : graph[rem])
                            q.push(nbr);
                    }
                    g = 1 - g;
                }
            }
        }
        return true;
    }
};

File: April 2022/April_29.java
class Solution {
    public boolean isBipartite(int[][] graph) {
        int n = graph.length;
        boolean[] vis = new boolean[n];
        int[] grp = new int[n];
        for(int i = 0; i < n; i++){
            if(vis[i])
                continue;
            Queue<Integer> q = new ArrayDeque<>();
            q.add(i);
            int g = 0;
            while(!q.isEmpty()){
                int size = q.size();
                while(size-- > 0){
                    int rem = q.remove();
                    if(vis[rem]){
                        if(grp[rem] == g)
                            continue;
                        return false;
                    }
                    vis[rem] = true;
                    grp[rem] = g;
                    
                    for(int nbr : graph[rem])
                        q.add(nbr);
                }
                g = 1 - g;
            }
        }
        return true;
    }
}
